smart-contracts/contracts/cl-vault/src/vault/deposit.rs

```diff
@@ -46,3 +68,523 @@ pub(crate) fn execute_deposit(
     // // Merge responses and add message to mint vault token
     // Ok(merge_responses(vec![receive_res, compound_res, mint_res]).add_event(event))
 }
+
+pub(crate) fn execute_multi_deposit(
+    deps: DepsMut,
+    env: Env,
+    info: &MessageInfo,
+    recipient: Option<String>,
+) -> Result<Response, ContractError> {
+    // Unwrap recipient or use caller's address
+    let recipient = recipient.map_or(Ok(info.sender.clone()), |x| deps.api.addr_validate(&x))?;
+
+    let pool = POOL_CONFIG.load(deps.storage)?;
+    let (token0, token1) = must_pay_two(&info, (pool.token0, pool.token1))?;
+
+    let position = POSITION.load(deps.storage)?;
+    let range = ConcentratedliquidityQuerier::new(&deps.querier)
+        .position_by_id(position.position_id)?
+        .position
+        .ok_or(ContractError::PositionNotFound)?
+        .position
+        .ok_or(ContractError::PositionNotFound)?;
+
+    let create_msg = create_position(
+        deps.storage,
+        &env,
+        range.lower_tick,
+        range.upper_tick,
+        vec![token0.clone(), token1.clone()],
+        Uint128::zero(),
+        Uint128::zero(),
+    )?;
+
+    CURRENT_DEPOSIT.save(
+        deps.storage,
+        &CurrentDeposit {
+            token0_in: token0.amount,
+            token1_in: token1.amount,
+            sender: recipient,
+        },
+    )?;
+
+    Ok(Response::new().add_submessage(SubMsg::reply_on_success(
+        create_msg,
+        Replies::DepositCreatePosition as u64,
+    )))
+}
+
+pub fn handle_deposit_create_position(
+    deps: DepsMut,
+    env: Env,
+    data: SubMsgResult,
+) -> ContractResult<Response> {
+    let resp: MsgCreatePositionResponse = data.try_into()?;
+    let current_deposit = CURRENT_DEPOSIT.load(deps.storage)?;
+    let bq = BankQuerier::new(&deps.querier);
+    let vault_denom = VAULT_DENOM.load(deps.storage)?;
+
+    // we mint shares according to the liquidity created
+    let liquidity = Decimal::from_str(resp.liquidity_created.as_str())?;
+
+    let total_position = get_position(deps.storage, &deps.querier, &env)?
+        .position
+        .ok_or(ContractError::PositionNotFound)?;
+    let total_liquidity = Decimal::from_str(total_position.liquidity.as_str())?;
+
+    // TODO change error type to something more descriptive
+    let total_shares: Uint128 = bq
+        .supply_of(vault_denom.clone())?
+        .amount
+        .ok_or(ContractError::IncorrectShares)?
+        .amount
+        .parse::<u128>()?
+        .into();
+
+    let ratio = liquidity.checked_div(total_liquidity)?;
+    let user_shares = total_shares.multiply_ratio(ratio.numerator(), ratio.denominator());
```

Comment:
can think about  possibly multiplying first and then divving. but minor -> leave to halborn